src/main/java/com/squareup/javapoet/ClassName.java

```diff
@@ -32,232 +31,106 @@
-      simpleNames.add(simpleName);
-      return simpleNames;
-    }
-
-    @Override
-    public ClassName peerClass(String name) {
-      return enclosingClassName.nestedClass(name);
-    }
-
-    @Override
-    ClassName prefixWithAtMostOneAnnotatedClass() {
-      if (hasAnnotatedEnclosingClass()) {
-        enclosingClassName.prefixWithAtMostOneAnnotatedClass();
-      }
+public final class ClassName extends TypeName implements Comparable<ClassName> {
+  public static final ClassName OBJECT = ClassName.get(Object.class);
 
-      return this;
-    }
+  /** The package name of this class, or "" if this is in the default package. */
+  final String packageName;
 
-    @Override
-    CodeWriter emitWithoutPrefix(
-        CodeWriter out, ClassName unannotatedPrefix) throws IOException {
+  /** The enclosing class, or null if this is not enclosed in another class. */
+  final ClassName enclosingClassName;
 
-      if (unannotatedPrefix.equals(this)) {
-        return out;
-      }
+  /** This class name, like "Entry" for java.util.Map.Entry. */
+  final String simpleName;
 
-      enclosingClassName.emitWithoutPrefix(out, unannotatedPrefix);
-      out.emit(".");
-      if (isAnnotated()) {
-        out.emit(" ");
-        emitAnnotations(out);
-      }
-      return out.emit(simpleName);
-    }
+  /** The full class name like "java.util.Map.Entry". */
+  final String canonicalName;
 
-    @Override
-    boolean hasAnnotatedEnclosingClass() {
-      return enclosingClassName.isAnnotated() || enclosingClassName.hasAnnotatedEnclosingClass();
-    }
+  private ClassName(String packageName, ClassName enclosingClassName, String simpleName) {
+    this(packageName, enclosingClassName, simpleName, Collections.emptyList());
   }
 
-  public static final ClassName OBJECT = ClassName.get(Object.class);
-
-  private ClassName(String simpleName, List<AnnotationSpec> annotations) {
+  private ClassName(String packageName, ClassName enclosingClassName, String simpleName,
+      List<AnnotationSpec> annotations) {
     super(annotations);
-    checkArgument(SourceVersion.isName(simpleName), "part '%s' is keyword", simpleName);
+    this.packageName = packageName;
+    this.enclosingClassName = enclosingClassName;
     this.simpleName = simpleName;
+    this.canonicalName = enclosingClassName != null
+        ? (enclosingClassName.canonicalName + '.' + simpleName)
+        : (packageName.isEmpty() ? simpleName : packageName + '.' + simpleName);
+  }
+
+  @Override public ClassName annotated(List<AnnotationSpec> annotations) {
+    return new ClassName(packageName, enclosingClassName, simpleName,
+        concatAnnotations(annotations));
   }
 
+  @Override public ClassName withoutAnnotations() {
+    ClassName resultEnclosingClassName = enclosingClassName != null
+        ? enclosingClassName.withoutAnnotations()
+        : null;
+    return isAnnotated()
+        ? new ClassName(packageName, resultEnclosingClassName, simpleName)
+        : this;
```

Comment:
I'm not sure this ternary should exist. What does `TypeName.get(Map.class).annotatedWith(..).nestedClass("Entry").withoutAnnotations()` produce?

Reply:
Just saving allocations when it's a no-op!

Reply:
Gotta get back in a Java desktop mood. This doesn't need every allocation to be justified.

Reply:
Ahhh, I see so it's also incorrect! Good catch!

Reply:
Or, put another way, this is equivalent to
```java
if (!isAnnotated()) return this;
ClassName resultEnclosingClassName = enclosingClassName != null
    ? enclosingClassName.withoutAnnotations()
    : null;
return new ClassName(packageName, resultEncosingClassName, simpleName)
```
Does `isAnnotated()` recurse on `enclosingClassName` for annotations?

Reply:
It doesn’t. I’ll fix.